File: backend/app/models/viagem_model.py
from enum import Enum

from sqlalchemy import Column, Integer, Date, Float, Text, Table, ForeignKey
from sqlalchemy.orm import relationship, Mapped, mapped_column

from app.core.postgres import Base

destino_viagens = Table(
    "destino_viagens",
    Base.metadata,
    Column("destino_id", Integer, ForeignKey("destino.id"), primary_key=True),
    Column("viagem_id", Integer, ForeignKey("viagem.id"), primary_key=True),
    extend_existing=True
)


class StatusViagem(Enum):
    PLANNING = "PLANNING"
    CONFIRMED = "CONFIRMED"
    ONGOING = "ONGOING"
    CANCELLED = "CANCELLED"


class ViagemModel(Base):
    __tablename__ = "viagem"

    id = Column(Integer, primary_key=True)
    data_inicio = Column(Date, nullable=False)
    data_fim = Column(Date, nullable=False)
    orcamento = Column(Float, nullable=False)
    observacao = Column(Text)
    descricao = Column(Text)
    status: Mapped[StatusViagem] = mapped_column(nullable=False)
    imagem = Column(Text)

    destinos = relationship("DestinoModel", secondary=destino_viagens)
    gastos = relationship("GastoModel", back_populates="viagem")

File: frontend/src/models/viagem.model.ts

export interface Tag {
    nome: string;
    descricao: string;
}

export interface Destino {
    id: number;
    nome: string;
    descricao: string;
    tags: Tag[];
}

export interface Gasto {
    id: number;
    descricao: string;
    valor: number;
    pago: boolean;
}

export interface Atividade {
    id: number;
    descricao: string
}

export interface DiaViagemAtividade {
    id: number;
    data_inicio: Date;
    data_fim: Date;
    custo: number;
    atividade: Atividade;
}

export interface DiaViagem {
    id: number;
    data_dia: Date;
    atividades: DiaViagemAtividade[];
}

export class Viagem {
    id: number;
    data_inicio: Date;
    data_fim: Date;
    orcamento: number;
    descricao: string;
    observacao: string;
    status: 'PLANNING' | 'CONFIRMED' | 'ONGOING' | 'CANCELLED';
    gastos?: Gasto[];
    dias?: DiaViagem[];
    imagem: string;
    destinos?: Destino[];

    constructor(data: any) {
        this.id = data.id;
        this.data_inicio = new Date(data.data_inicio);
        this.data_fim = new Date(data.data_fim);
        this.orcamento = data.orcamento;
        this.descricao = data.descricao;
        this.observacao = data.observacao;
        this.status = data.status;
        this.gastos = data.gastos;
        this.dias = data.dias;
        this.imagem = data.imagem;
        this.destinos = data.destinos;
    }

}